src/shogun/regression/GLM.h

```diff
@@ -0,0 +1,112 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Ahmed Khalifa
+ */
+/*
+References:
+https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/family
+https://www.statsmodels.org/stable/generated/statsmodels.genmod.generalized_linear_model.GLM.html#statsmodels.genmod.generalized_linear_model.GLM
+http://glm-tools.github.io/pyglmnet/api.html
+*/
+#ifndef _GENERALIZEDLINEARMODEL_H__
+#define _GENERALIZEDLINEARMODEL_H__
+
+#include <shogun/lib/config.h>
+
+#include <shogun/features/Features.h>
+#include <shogun/machine/FeatureDispatchCRTP.h>
+#include <shogun/machine/LinearMachine.h>
+#include <shogun/optimization/DescendUpdater.h>
+#include <shogun/regression/Regression.h>
+
+namespace shogun
+{
+	enum Family
+	{
+		NORMAL_DISTRIBUTION,
+		EXPONENTIAL_DISTRIBUTION,
+		GAMMA_DISTRIBUTION,
+		BINOMIAL_DISTRIBUTION,
+		GAUSS_DISTRIBUTION,
+		POISSON_DISTRIBUTION
+	};
+	enum LinkFunction
+	{
+		LOG,
+		LOGIT,
+		IDENTITY,
+		INVERSE
+	};
+	/** @brief Class GLM implements Generalized Linear Models, such as poisson,
+	 * gamma, binomial
+	 */
+	class GLM : public LinearMachine
+	{
+	public:
+		/** problem type */
+		MACHINE_PROBLEM_TYPE(PT_REGRESSION);
+
+		/** Default constructor */
+		GLM();
+
+		/** Constructor
+		 *
+		 * @param descend_updater chosen Descend Updater algorithm
+		 * @param Linkfn the link function check
+		 * https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/family
+		 * @param Family the family check
+		 * https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/family
+		 * @param alpha alpha
+		 * @param lambda lambda
+		 */
+		GLM(const std::shared_ptr<DescendUpdater>& descend_updater,
```

Comment:
could you remove those ctors? The new API design uses the default ctor with subsequent `put` calls

Reply:
in SWIG... i would still have this for c++

Reply:
fair enough. in that case, please make sure the class works with the default ctor. We had/have many bugs where a class segfaults in the train method as the default ctor didnt initialize things and the train method didnt check that. Don't rely that a user used this ctor, write separate tests for both

Reply:
and also please make sure that the default parameters in the ctor match the default parameters when using the default one (another source of confusion). I would remove the default values in the ctor to be on the safe side actually, and instead you can initialize the values in the `::init` method

Reply:
You can even have the values inline with the class member declaration